File: lib/seriport.cc
/*
 * inci driver
 * Copyright (C) 2008 i Bilisim Teknolojileri Ltd. Sti.
 *
 */

#include <fcntl.h>
#include <termios.h>
namespace unistd {
#include <unistd.h>
};

#include "seriport.h"

static speed_t
get_speed_enum(unsigned int speed)
{
	if (speed == 1200) return B1200;
	if (speed == 2400) return B2400;
	if (speed == 4800) return B4800;
	if (speed == 9600) return B9600;
	if (speed == 19200) return B19200;
	if (speed == 38400) return B38400;
	if (speed == 57600) return B57600;
	if (speed == 115200) return B115200;

	// Unknown
	return B0;
}

SeriPort::SeriPort(const char *portdev, unsigned int speed)
{
	is_ok = 0;

	fd = open(portdev, O_RDWR);
	if (fd == -1) return;

	struct termios prefs;

	if (0 != tcgetattr(fd, &prefs)) return;

	// Set speed
	speed_t st = get_speed_enum(speed);
	if (st == B0) return;
	if (0 != cfsetispeed(&prefs, st)) return;
	if (0 != cfsetospeed(&prefs, st)) return;

	// Activate receiver
	prefs.c_cflag |= (CLOCAL | CREAD);

	// Disable echo, output modification, etc
	prefs.c_lflag = 0;
	prefs.c_oflag = 0;

	// Set mode (8n1)
	prefs.c_cflag &= ~CSIZE; // clear mode bits first
	prefs.c_cflag |= CS8;
	prefs.c_cflag &= ~(PARENB | PARODD);
	prefs.c_cflag &= ~CSTOPB;

	// No flow control
	prefs.c_iflag &= ~(IXON | IXOFF | IXANY);
	prefs.c_cflag &= ~CRTSCTS;

	// Dont mess with CRLF chars
	prefs.c_iflag &= ~(IGNCR | ICRNL | INLCR);

	// Ignore parity and break
	prefs.c_iflag |= IGNPAR | IGNBRK;

	if (0 != tcsetattr(fd, TCSANOW, &prefs)) return;

	is_ok = 1;
}

SeriPort::~SeriPort()
{
	if (fd != -1) unistd::close(fd);
}

int
SeriPort::read(char *buffer, size_t len)
{
	return unistd::read(fd, (void *) buffer, len);
}

int
SeriPort::write(const char *buffer, size_t len)
{
	return unistd::write(fd, (void *) buffer, len);
}

void
SeriPort::flush_input()
{
	tcflush(fd, TCIFLUSH);
}

File: python/seriport.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys
import os
import fcntl
import termios
import struct
import time

IFLAG, OFLAG, CFLAG, LFLAG, ISPEED, OSPEED, CC = range(7)

def list_ports():
    # This doesn't work with 2008 kernel
    return map(lambda x: os.path.realpath("/dev/tts/" + x), os.listdir("/dev/tts"))


class Port:
    def __init__(self, device_name, speed=38400, mode="8n1"):
        if mode != "8n1":
            raise ValueError("Mode '%s' is not yet supported" % mode)
        
        self.fd = os.open(device_name, os.O_RDWR)
        prefs = termios.tcgetattr(self.fd)
        
        # Set speed
        speed_val = getattr(termios, "B" + str(speed), None)
        if not speed_val:
            raise ValueError("Unknown speed value '%s'" % speed)
        prefs[OSPEED] = prefs[ISPEED] = speed_val
        
        # Activate receiver
        prefs[CFLAG] |= (termios.CLOCAL | termios.CREAD)
        
        # Disable echo, output modification, etc
        prefs[LFLAG] = 0
        prefs[OFLAG] = 0
        
        # Set mode (8n1)
        prefs[CFLAG] &= ~termios.CSIZE
        prefs[CFLAG] |= termios.CS8
        prefs[CFLAG] &= ~(termios.PARENB|termios.PARODD)
        prefs[CFLAG] &= ~termios.CSTOPB
        
        # No flow control
        prefs[IFLAG] &= ~(termios.IXON|termios.IXOFF | termios.IXANY)
        prefs[CFLAG] &= ~(termios.CRTSCTS)
        
        # Dont mess with CRLF chars
        prefs[IFLAG] &= ~(termios.IGNCR | termios.ICRNL | termios.INLCR)
        
        # Ignore parity and break
        prefs[IFLAG] |= termios.IGNPAR | termios.IGNBRK
        
        termios.tcsetattr(self.fd, termios.TCSANOW, prefs)
    
    def __getattr__(self, name):
        if name == "cts":
            s = fcntl.ioctl(self.fd, termios.TIOCMGET, struct.pack('I', 0))
            return struct.unpack('I',s)[0] & termios.TIOCM_CTS != 0
        elif name == "dsr":
            s = fcntl.ioctl(self.fd, termios.TIOCMGET, struct.pack('I', 0))
            return struct.unpack('I',s)[0] & termios.TIOCM_DSR != 0
        elif name == "ri":
            s = fcntl.ioctl(self.fd, termios.TIOCMGET, struct.pack('I', 0))
            return struct.unpack('I',s)[0] & termios.TIOCM_RI != 0
        elif name == "cd":
            s = fcntl.ioctl(self.fd, termios.TIOCMGET, struct.pack('I', 0))
            return struct.unpack('I',s)[0] & termios.TIOCM_CD != 0
        else:
            raise AttributeError, name
    
    def __setattr__(self, name, value):
        if name == "rts":
            cmd = termios.TIOCMBIC
            if value:
                cmd = termios.TIOCMBIS
            fcntl.ioctl(self.fd, cmd, struct.pack('I', termios.TIOCM_RTS))
        elif name == "dtr":
            cmd = termios.TIOCMBIC
            if value:
                cmd = termios.TIOCMBIS
            fcntl.ioctl(self.fd, cmd, struct.pack('I', termios.TIOCM_DTR))
        else:
            self.__dict__[name] = value
    
    def read(self, len):
        return os.read(self.fd, len)
    
    def write(self, data):
        return os.write(self.fd, data)
    
    def flush_input(self):
        termios.tcflush(self.fd, termios.TCIFLUSH)


if __name__ == "__main__":
    print "Detected devices:", list_ports()
    port = Port("/dev/ttyUSB0")
    port = Port("/dev/ttyUSB0", speed=115200)
    print port.cts, port.dsr, port.ri, port.cd
    #port.rts = True
    #port.rts = False
    #port.dtr = True
    #port.dtr = False
    port.write("lala")
    print port.read(10)
